### PR TITLE
New Chroma Controller GUI and other visual improvements

### PR DESCRIPTION
**Pull request by lah7**
First of all, thank you very much to all the contributors who put this driver together! Razer, if you're reading this, I wouldn't have bought this awesome keyboard if it wasn't for these open source drivers!

**Here's an overview of my code contributions:**

### GUI "Controller" Application
 * Based on Python3 and WebKit.
 * Same features as the Tray Applet.
 * New 'basic' profile support (using 'dynamic') - creation, edit static colour keys, deletion.
 * Dynamic Profiles are saved to `~/.config/razer_chroma/profiles`
 * Currently only has English UK layout, but more can be added to the `data/layouts/` folder as custom HTML files.
 * If the default "flowing colours" effect from the daemon is running, a different mode needs to be selected first.

### Tray Applet
  * Add colour picker so the colour of the static, reactive and breath effects can be quickly changed.

### Sharing Code
Code for communicating with the dbus service has been moved into its own file for importing. I haven't quite figured out a way to do this properly. For now, they expect the file to be in the same folder, so a symbolic link will be required for the GUI Controller:

    ln -s gui/tray_applet/daemon_dbus.py gui/chroma_controller/daemon_dbus.py

The installation copies this file automatically.

### Other Changes / Tweaks
  * The package "python-appindicator" was missing, but required for the Tray Applet.
  * Fix the launcher not appearing on the MATE desktop environment.
  * Freshen up the README's styling, plus add details about the GUIs.
  * Installation scripts updated for new GUIs.
  * Created a new icon to represent the drivers. A flat keyboard inspired by Razer's snakes.

All changes work as intended under Ubuntu MATE 15.10 amd64.

### Upcoming Changes
You may notice a few FIXMEs. I'm willing to keep this GUI maintained as new commits improve the backend of things and send pull requests when the GUI is stable enough for general usage. Some planned changes include:
  * Check if the Chroma Keyboard is plugged in / the DBUS service is running.
  * **Adding preferences:**
   * Change keyboard locale layout.
   * Option to instantly activate profiles when switching.
   * Option to see profile changes live while editing.
   * Optionally set a starting colour.
  * Switch profiles from the tray.
  * Try to retain the colour when opening the colour picker.
  * Python script that easily imports `daemon_dbus.py` for controlling via bash scripts.

### Changes that affect things...
**Needs decisions/comments on:**
  * Which way is best to share 'libraries' between the Tray/GUI.
   * Move both GUIs to own folder (eg. `/usr/share/razer_chroma_driver`)?
   * Merge both GUIs (Controller + Tray Applet) together?

### To make things even better...
The following would be needed:
 1. ~~Fix the 'Breath' effect.~~ (Fixed in #44 )
 2. Ability to restore original 'flowing colours' effect (the one used when `razer_bcd` service first starts? Not sure what it's called?)
 3. Merging `dynamic` within daemon (dbus) to prevent need for root?
 4. `dynamic` had the ability to bind 'effects' to specific keys, eg: "Wave" for F1-F12.

Some of these ideas to code go beyond my knowledge.

--------
Hope you like these new additions.

### CODE REVIEW

**Comment by terrycain** [comment]
Am a bit busy at the moment. But I have a load of improvements for the kernel driver (for one, properly fixing breath mode) and I have a Razer Mamba driver to add.

File: gui/chroma_controller/chroma_controller.py [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+import daemon_dbus
+
+# Default Settings & Preferences
+rgb_effects_red = 0;
+rgb_effects_green = 255;
+rgb_effects_blue = 0;
+current_effect = 'custom';
+layout = 'en-gb';
+
+class Paths():
```
Comment (terrycain) [inline gui/chroma_controller/chroma_controller.py]: should be class `Paths(object):` ;-)

Comment (lah7) [inline gui/chroma_controller/chroma_controller.py]: Thanks for spotting that - amended to last commit.

**Comment by lah7** [comment]
**Note to maintainer:** Consider reviewing #44 first, as I'll update the GUIs to reflect the new fixes and update this request as soon as possible.

**Comment by pez2001** [comment]
got a merging conflict , i guess you were expecting this... i will look into it 
bg
tim

**Comment by lah7** [comment]
Don't worry, I need to make some minor amendments first due to these new changes.

**Comment by terrycain** [comment]
> Merging dynamic within daemon (dbus) to prevent need for root?

As for the dynamic binary, we could probably add support for the  set_key_row file on the daemon and then you could interact with that using DBUS. Just trying to think of the best way to do that.

---
> Which way is best to share 'libraries' between the Tray/GUI.

The best way would be to make a module called something `razer_lib` and stick it in `/usr/lib/python3/dist-packages` then it'll be in the path and can be imported from anywhere.

---
> Python script that easily imports daemon_dbus.py for controlling via bash scripts

What are you trying to accomplish with this as if your doing what I think your doing, I have a few ideas.

---
> Ability to restore original 'flowing colours' effect (the one used when razer_bcd service first starts? Not sure what it's called?)

Its to do with the daemon, When you set one of the core effects it will pause the effect render loop in the daemon. In theory if we just unpause it the current effect in the daemon should continue.

---
> dynamic had the ability to bind 'effects' to specific keys, eg: "Wave" for F1-F12.

Do you mean set the wave effect for just the F1-F12 keys or do you mean get the F1-F12 keys activate different effects for the whole of the keyboard?

**Comment by lah7** [comment]
Thanks @terrycain for addressing these.

-----
> As for the dynamic binary, we could probably add support for the set_key_row file on the daemon and then you could interact with that using DBUS. Just trying to think of the best way to do that.

Forgive my misunderstanding, would that be to set a key at (row, column) (x,y) or do you mean setting the entire row?

------
> Python script that easily imports daemon_dbus.py for controlling via bash scripts

By this, I'm thinking along the lines of a "bridge" between Bash to Python's DBUS *(being a reliable way to control keyboard without needing root)*. Such options could be switching profiles or to control the keyboard by commands. For instance, from the terminal and bash scripts:
     `chroma effect wave`
     `chroma profile name_of_game`
     `chroma brightness off`

It's an alternate interface then the GUIs and without needing to be root - automation would be easier to integrate into existing scripts for instance, or to bind as keyboard shortcuts within the desktop environment if desired.

----
> Do you mean set the wave effect for just the F1-F12 keys

Yes, similar to how Synapse can configure keys to use different colours and effects. I'm wondering if it's possible like so:
`2 1 1 1 2 1 3 1 4 1 5`
`<2 = Set Wave> <x1> <y1> [x2] [y2] [x3] [y3]`
Similar to:
 `1 3 2 0 255 0 `
`<1 = Set key> <x> <y> <red> <green> <blue>`

**Comment by lah7** [comment]
The conflicts are caused by `gui/tray_applet/razer_tray_applet.py` whereby these changes are mostly migrated to `gui/tray_applet/daemon_dbus.py`.

Something is going horribly wrong in my local forked repo. It would be safer if you proceed to merge and either:
 * Overwrite, using my version of `razer_tray_applet.py` for now, then I'll amend all necessary changes in the next pull request.
 * Migrate changes from `razer_tray_applet.py` to `daemon_dbus.py` where applicable.

Sorry to be inconvenient, I'm more familiar with BitBucket then GitHub.

**Comment by terrycain** [comment]
> Forgive my misunderstanding, would that be to set a key at (row, column) (x,y) or do you mean setting the entire row?

Well at a hardware level to set any key, the entire row must be set even if there is no change to most of the keys. From a design point of view we could do something like the scripts that `dynamic` uses (but a bit more verbose so its easier to understand). E.g. Read in a keyboard colour config and then send to dbus what the whole keyboard will look like. You with me yet?

---
> By this, I'm thinking along the lines of a "bridge" between Bash to Python's DBUS

Well have you looked at `razer_bcd_controller` that does roughly what you want but I find its quite buggy. I was planning on making one in Python as then I can parse the arguments better and the DBUS is slightly easier. Plus it'll need to be extended to allow Mouse and eventually firefly support.

---
> Yes, similar to how Synapse can configure keys to use different colours and effects. I'm wondering if it's possible like so:

Its not impossible but doing anything other than the base effects on the keyboard require loading animations into the daemon etc...

---

I'll have a look at the merge in a bit.

**Comment by terrycain** [comment]
This link "should" give you a git repo that I've edited. [Git link](http://s000.tinyupload.com/index.php?file_id=00895424416331866228)

**Comment by lah7** [comment]
Thanks for that. For some reason, I can't recompile:
`Makefile:9: recipe for target 'razer_kbd' failed`
`make: *** [razer_kbd] Error 2`

That's what happened on my attempt, so I'm not sure what's changed to cause that. A fresh clone of this repository compiles just fine...?

**Pull request merged**